File: rsa.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function to check if a number is prime or not
bool isPrime(int n) {
    if (n <= 1) {
        return false;
    }
    for (int i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}

// Function to find GCD of two numbers
int gcd(int a, int b) {
    if (b == 0) {
        return a;
    }
    return gcd(b, a % b);
}

int main() {
    // Step 1: Choose two prime numbers
    int p = 17, q = 11;

    // Step 2: Compute n and phi
    int n = p * q;
    int phi = (p - 1) * (q - 1);

    // Step 3: Choose an integer e such that 1 < e < phi and gcd(e, phi) = 1
    int e = 7;
    while (e < phi) {
        if (gcd(e, phi) == 1) {
            break;
        }
        e++;
    }

    // Step 4: Compute the secret key d
    int k = 2;
    int d = (1 + (k * phi)) / e;
    while (d > phi) {
        k++;
        d = (1 + (k * phi)) / e;
    }

    // Step 5: Print the public and private keys
    cout << "Public key: {" << e << ", " << n << "}" << endl;
    cout << "Private key: {" << d << ", " << n << "}" << endl;

    // Step 6: Encrypt the message
    string message = "HELLO";
    int encrypted[message.length()];
    for (int i = 0; i < message.length(); i++) {
        int m = message[i];
        int c = fmod(pow(m, e), n);
        encrypted[i] = c;
    }

    // Step 7: Decrypt the message
    string decrypted;
    for (int i = 0; i < message.length(); i++) {
        int c = encrypted[i];
        int m = fmod(pow(c, d), n);
        decrypted += (char)m;
    }

    // Step 8: Print the encrypted and decrypted messages
    cout << "Encrypted message: ";
    for (int i = 0; i < message.length(); i++) {
        cout << encrypted[i] << " ";
    }
    cout << endl;
    cout << "Decrypted message: " << decrypted << endl;

    return 0;
}

File: rsa.py
import math
import random 

# Calculate GCD using Euclid's Algorithm
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# Check if a number is prime
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

# Generate public and private keys
def generate_keypair(p, q):
    if not (is_prime(p) and is_prime(q)):
        raise ValueError("Both numbers must be prime.")
    n = p * q
    phi = (p - 1) * (q - 1)
    # Choose e such that 1 < e < phi and e is coprime to phi
    e = random.randrange(1, phi)
    while gcd(e, phi) != 1:
        e = random.randrange(1, phi)
    # Calculate d such that ed = 1 (mod phi)
    d = pow(e, -1, phi)
    return ((n, e), (n, d))

# Encrypt a message
def encrypt(pk, plaintext):
    n, e = pk
    # Convert each letter in the plaintext to numbers based on the character using a^b mod m
    cipher = [pow(ord(char), e, n) for char in plaintext]
    return cipher

# Decrypt a message
def decrypt(pk, ciphertext):
    n, d = pk
    # Generate the plaintext based on the ciphertext and key using a^b mod m
    plain = [chr(pow(char, d, n)) for char in ciphertext]
    return ''.join(plain)

# Example usage
p = 17
q = 19
public_key, private_key = generate_keypair(p, q)
print("Public key:", public_key)
print("Private key:", private_key)
message = "Mihir"
encrypted_message = encrypt(public_key, message)
print("Encrypted message:", encrypted_message)
decrypted_message = decrypt(private_key, encrypted_message)
print("Decrypted message:", decrypted_message)